x-pack/plugins/ml/public/data_frame/components/define_pivot/define_pivot_form.tsx

```diff
@@ -121,48 +127,21 @@ export const DefinePivotForm: SFC<Props> = React.memo(({ overrides = {}, onChang
     setAggList(newList);
   };
 
-  // The available fields for group by
-  const groupByOptions: EuiComboBoxOptionProps[] = [];
-  fields.forEach(field => {
-    const o: DropDownLabel = { label: field.name };
-    groupByOptions.push(o);
-  });
-
-  // The available aggregations
-  const aggOptions: EuiComboBoxOptionProps[] = [];
-  const aggOptionsData: Dictionary<OptionsDataElement> = {};
-
-  fields.forEach(field => {
-    const o: DropDownOption = { label: field.name, options: [] };
-    pivotSupportedAggs.forEach(agg => {
-      if (
-        (agg === PIVOT_SUPPORTED_AGGS.CARDINALITY &&
-          (field.type === FIELD_TYPE.STRING || field.type === FIELD_TYPE.IP)) ||
-        (agg !== PIVOT_SUPPORTED_AGGS.CARDINALITY && field.type === FIELD_TYPE.NUMBER)
-      ) {
-        const label = `${agg}(${field.name})`;
-        o.options.push({ label });
-        const formRowLabel = `${agg}_${field.name}`;
-        aggOptionsData[label] = { agg, field: field.name, formRowLabel };
-      }
-    });
-    aggOptions.push(o);
-  });
-
   const pivotAggs = aggList.map(l => aggOptionsData[l]);
-  const pivotGroupBy = groupBy;
+  const pivotGroupByArr = dictionaryToArray(groupByList);
   const pivotQuery = getPivotQuery(search);
 
-  const valid = pivotGroupBy.length > 0 && aggList.length > 0;
-
+  const valid = pivotGroupByArr.length > 0 && aggList.length > 0;
   useEffect(
     () => {
-      onChange({ aggList, aggs: pivotAggs, groupBy: pivotGroupBy, search, valid });
+      onChange({ aggList, aggs: pivotAggs, groupByList, search, valid });
     },
     [
       aggList,
       pivotAggs.map(d => `${d.agg} ${d.field} ${d.formRowLabel}`).join(' '),
-      pivotGroupBy,
+      pivotGroupByArr
+        .map(d => `${d.agg} ${d.field} ${'interval' in d ? d.interval : ''} ${d.formRowLabel}`)
```

Comment:
Just want to make sure it feels safe to use the `in` operator here since it matches all object keys, including ones in the object's prototype chain. Is there any possibility there could be a similar key that could return a false positive?

Reply:
Thanks for pointing this out! I initially tried using `hasOwnProperty()` but that wouldn't work in place with TypeScript so I ended up using this version. I now changed the code to use a type guard which uses `hasOwnProperty()` internally in fe27ec7.